### PR TITLE
nixos/security/vault: init

### PR DESCRIPTION
**Pull request by RealityAnomaly**
###### Motivation for this change
This pull request adds a module for dynamic declarative deployment of HashiCorp Vault keys as well as a module to configure and start `vault-agent` as a service. Keys are configured as templates in the agent which allows for automatic renewal.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
cc vault maintainers @rushmorem @lnl7 @offlinehacker @pradeepchhetri @Chili-Man 

Nitpick: this adds `services.vault-agent` whereas the commit and PR call it `vault`. Maybe change that to `vault-agent`?

**Comment by CitadelCore** [comment]
> cc vault maintainers @rushmorem @LnL7 @offlinehacker @pradeepchhetri @Chili-Man
> 
> Nitpick: this adds `services.vault-agent` whereas the commit and PR call it `vault`. Maybe change that to `vault-agent`?

Good point with the naming, I should probably split up the commits anyway - one to add the agent, the other to add the keys module.

**Comment by CitadelCore** [comment]
Commits have been split up.

**Comment by jonringer** [comment]
cc @manveru

**Comment by jonringer** [comment]
Still relevant, just a lot to review

**Review by bobbbay — COMMENTED** [review]
My (literally) two cents:

File: nixos/modules/security/vault/keys.nix [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+
+        field = mkOption {
+          default = null;
+          type = types.nullOr types.str;
+          description = "Default key/value field to use.";
+        };
+      };
+    });
+
+    pki = types.submodule ({ config, ... }: {
```
Comment (RaitoBezarius) [inline nixos/modules/security/vault/keys.nix:101]: It does not seem you need `{ config, ... }:` here.

File: nixos/modules/security/vault/keys.nix [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+    security.vault-keys = mkOption {
+      default = { };
+      type = types.attrsOf keyType;
+      description = "Attributes of vault keys to deploy.";
+    };
+
+    reflection.vault-keys = mkOption {
+      default = finalKeys;
+      type = types.attrs;
+      description = "Allows extration of template values by an external script.";
```
Comment (RaitoBezarius) [inline nixos/modules/security/vault/keys.nix:275]: extraction

File: nixos/modules/security/vault/keys.nix [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+  options = {
+    security.vault-keys = mkOption {
+      default = { };
+      type = types.attrsOf keyType;
+      description = "Attributes of vault keys to deploy.";
+    };
+
+    reflection.vault-keys = mkOption {
+      default = finalKeys;
+      type = types.attrs;
```
Comment (RaitoBezarius) [inline nixos/modules/security/vault/keys.nix:274]: `types.attrs` is deprecated: `types.attrsOf types.anything` is the replacement, I believe.
If the value is a script, we could do `types.attrsOf (types.oneOf [ types.package types.path ])` I think?

File: nixos/modules/security/vault/keys.nix [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+
+              # if the key's folder does not exist, create it
+              if [[ ! -d "${key.folder}" ]]; then
+                  mkdir -p "${key.folder}"
+                  chmod -R 0600 "${key.folder}"
+              fi
+
+              # we are now DONE waiting for all the keys
+              # finally render them
+            ${concatStringsSep "\n" (flatten ([
```
Comment (RaitoBezarius) [inline nixos/modules/security/vault/keys.nix:343]: Some indent problem here

File: nixos/modules/security/vault/keys.nix [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+              (flip mapAttrsToList key.sinks (_: sink:
+              let
+              buildKey = render: dest: id:
```
Comment (RaitoBezarius) [inline nixos/modules/security/vault/keys.nix:346]: It would be nice to indent one level

File: nixos/modules/security/vault/keys.nix [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+                    ${render}
+                    chmod ${sink.permissions} "${dest}"
+                    chown "${sink.user}:${sink.group}" "${dest}"
+                fi
+              '';
+
+              buildStandaloneKey =
+              let
+                  dest = "${key.folder}/${sink.name}";
+              in
```
Comment (RaitoBezarius) [inline nixos/modules/security/vault/keys.nix:363]: Why do you need this `in`?

Comment (RaitoBezarius) [reply nixos/modules/security/vault/keys.nix:363]: (I am trying to make sense of the whole order, it is a bit hard, I think, moving stuff in a library or encapsulating logic would help a lot here.)

File: nixos/modules/security/vault/keys.nix [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+                  buildKey ''cat "${key.folder}/.${default.id}.tmp" | ${pkgs.jq}/bin/jq -j -r '.["${sink.kv.field}"]' > "${dest}"'' dest default.id
+
+                  # fallback to regular
+                  else buildStandaloneKey))
+
+                  # remove the temporary files, but not for external keys as we might need to re-render them
+                  # TODO: could we actually just do this for all keys, i.e. not remove the tempfiles?
```
Comment (RaitoBezarius) [inline nixos/modules/security/vault/keys.nix:388]: what would bring to do this for all keys?

Comment (RaitoBezarius) [reply nixos/modules/security/vault/keys.nix:388]: can we write a NixOS test exerting an external key which require a rerender?

File: nixos/modules/services/security/vault-agent.nix [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+
+    ${optionalString cfg.cache.enable ''
+    cache {
+        use_auto_auth_token = ${toString cfg.cache.useAutoAuthToken}
+    }
+    ''}
+
+    vault {
+        address = "${cfg.address}"
+        ${if (cfg.caCertFile != null) then ''ca_cert = "${cfg.caCertFile}"'' else ''''}
```
Comment (RaitoBezarius) [inline nixos/modules/services/security/vault-agent.nix:36]: wouldn't `""` be better for the empty string?

File: nixos/modules/services/security/vault-agent.nix [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+          type = types.enum [ "tcp" "unix" ];
+          default = "tcp";
+          description = "Type of the listener to use";
+        };
+
+        address = mkOption {
+          type = types.str;
+          default = "127.0.0.1:8200";
+          description = "Address of the listener to listen to";
+        };
```
[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply nixos/modules/services/security/vault-agent.nix:99]: Wouldn't that be a runtime error?

File: nixos/modules/services/security/vault-agent.nix [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+rec {
+  options = {
+    services.vault-agent = {
+      enable = mkEnableOption "Vault agent";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.vault;
+        defaultText = "pkgs.vault";
+        description = "This option specifies the vault package to use.";
```
Comment (RaitoBezarius) [inline nixos/modules/services/security/vault-agent.nix:183]: `HashiCorp Vault` would be clearer I think

File: nixos/modules/services/security/vault-agent.nix [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+          default = null;
+          description = "Checks to ensure the path is within this path.";
+        };
+      };
+    };
+in
+rec {
+  options = {
+    services.vault-agent = {
+      enable = mkEnableOption "Vault agent";
```
Comment (RaitoBezarius) [inline nixos/modules/services/security/vault-agent.nix:177]: Nitpick: `HashiCorp Vault agent`

**Review by RaitoBezarius — COMMENTED** [review]
An amazing piece of code, I really want this now. :>
A suggestion to make it realistic to merge it for NixOS 22.11: wouldn't it be better to split vault-keys and vault-agent in two PRs?

I feel like `vault-agent` can be merged quite easily, `vault-keys` is a lot more involved IMHO.

**Comment by emilylange** [comment]
Closing in favor of #216230 which got merged a few weeks ago and added `services.vault-agent` and `services.consul-template` and is now available in 23.05

See https://github.com/NixOS/nixpkgs/blob/f29031e265170c9d67511a7834658e646f98d499/nixos/doc/manual/release-notes/rl-2305.section.md#new-services-sec-release-2305-new-services

**Pull request closed without merging**